src/Compilers/Core/Portable/DiagnosticAnalyzer/AsyncQueue.cs

```diff
@@ -94,9 +96,12 @@ private bool EnqueueCore(TElement value)
                 waiter = _waiters.Dequeue();
             }
 
-            // Invoke SetResult on a separate task, as this invocation could cause the underlying task to executing,
-            // which could be a long running operation that can potentially cause a deadlock if executed on the current thread.
-            Task.Run(() => waiter.SetResult(value));
+            Debug.Assert(waiter.Task.CreationOptions.HasFlag(TaskCreationOptions.RunContinuationsAsynchronously));
+            if (!waiter.TrySetResult(value))
+            {
+                // A waiter was available in the queue, but was cancelled before we were able to assign this value
+                goto retry;
```

Comment:
any concern that we might hit this, loop, and then hit the `if (_disallowEnqueue)` case?

Reply:
This can only occur for two unordered operations on two different threads, with one calling Enqueue and the other calling Complete. In this case, the caller would observe the Complete call occurring first.

Reply:
Agree on the impact but that also seems fine to me. If one thread is calling `Enqueue` and the other is calling `Complete` there is really no guarantee that `Enqueue` will complete (haha). It's a race.

Reply:
➡️ No action taken